app/controllers/import/uploads_controller.rb

```diff
@@ -0,0 +1,47 @@
+class Import::UploadsController < ApplicationController
+  layout "imports"
+
+  before_action :set_import
+
+  def show
+  end
+
+  def update
+    if csv_valid?(csv_str)
+      @import.assign_attributes(raw_file_str: csv_str, col_sep: upload_params[:col_sep])
+      @import.save!(validate: false)
+
+      redirect_to import_configuration_path(@import), notice: "CSV uploaded successfully."
+    else
+      flash.now[:alert] = "Must be valid CSV with headers and at least one row of data"
+
+      render :show, status: :unprocessable_entity
+    end
+  end
+
+  private
+    def set_import
+      @import = Current.family.imports.find(params[:import_id])
+    end
+
+    def csv_str
+      @csv_str ||= upload_params[:csv_file]&.read || upload_params[:raw_file_str]
+    end
+
+    def csv_valid?(str)
+      require "csv"
+
+      begin
+        csv = CSV.parse(str || "", headers: true)
+        return false if csv.headers.empty?
+        return false if csv.count == 0
+        true
+      rescue CSV::MalformedCSVError
+        false
+      end
+    end
```

Comment:
Similar to some of my comments above, I agree that there is an opportunity to clean this up a bit, but this controller and several of the other import step controllers are already pretty thin since I've split them into distinct steps.  

I'm just not seeing a huge benefit to adding more layers of abstraction to something this simple.